### ISSUE TITLE
Distribution lookup is case-sensitive

### ISSUE THREAD

**Issue by jaraco** [python/importlib_metadata#20]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 1, 2018, 20:12

```
~ $ rwt -q importlib_metadata==0.6 -- -c "import importlib_metadata; print(importlib_metadata.version('PyJWT'))"
1.6.4
~ $ rwt -q importlib_metadata==0.6 -- -c "import importlib_metadata; print(importlib_metadata.version('pyjwt'))"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-_mwk2v6o/importlib_metadata/api.py", line 105, in version
    return distribution(package).version
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-_mwk2v6o/importlib_metadata/api.py", line 86, in distribution
    return Distribution.from_name(package)
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/rwt-_mwk2v6o/importlib_metadata/api.py", line 52, in from_name
    raise PackageNotFoundError(name)
importlib_metadata.api.PackageNotFoundError: pyjwt
```

Probably the search should be case-insensitive.

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on Nov 2, 2018, 24:32

Do you know if it's possible to upload different packages to PyPI which differ only in their case?  I sure hope the answer is "no".

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 2, 2018, 24:51

Definitely no. The names are normalized by the packaging tools.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 2, 2018, 01:00

I was about to write a test to capture this issue when I realized the tests aren't running with pytest, so I don't get fixtures... so creating a test against which to validate is not fun when the tools I would use to do that are not present.

As time permits, I'll learn what facilities unittest provides.

I also wonder - should importlib_resources support all package name normalization (e.g. `_` -> `-`) or just case insensitivity?

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 2, 2018, 01:02

> should importlib_resources support all package name normalization (e.g. `_` -> `-`) or just case insensitivity?

On further consideration, probably not... because the names being used are not normalized at all... so the user can be expected to provide the canonical name, and we'll add case insensitivity for convenience.

**Comment by jaraco** [comment]
In GitLab by [@warsaw](https://gitlab.com/warsaw) on Nov 2, 2018, 01:10

I concur that we should keep package name normalization to a minimum.  IME, it can cause many problems, especially when competing systems are normalizing or not with subtly different rules.

I think case insensitivity makes sense, but I'd like to see the implementation.  Given that the name on the file system could differ by case, you'll have to be somewhat careful in how you match the file system names.  Maybe it won't be too difficult though.  We'll need a really good set of tests to cover the conditions.

As for `unittest` vs. `pytest` - the main reason for using `unittest` and keeping external dependencies to a minimum are because this package will have to get ported to Python 3.8.  If we use `pytest` here, that will be more difficult.  (The same holds true for `importlib_resources` of course.)

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 10, 2018, 16:22

mentioned in commit d4da27e9abeba6ded2d968b9b65ece15c0ef4c5d

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 10, 2018, 16:22

mentioned in commit 0e945cf6cc45c899329d692571963eed6a76de5f

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 10, 2018, 16:22

mentioned in commit 8bf077033ec6b3ce0e3d7ee51a90bb298ce701c1

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 10, 2018, 16:22

mentioned in commit a017b2d8d2f8f8499cbe5b78847767fb12039a26

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 10, 2018, 16:26

mentioned in merge request !27

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 10, 2018, 16:28

mentioned in commit cf550b3cd6f4c79854d2ff7e5a0329e896d4cccf

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 10, 2018, 19:33

mentioned in commit 4f67ef8c7c6a7e768df4c95f142e0972721f405d

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 10, 2018, 19:33

mentioned in commit ef0dc89d00d0be4e598a8d9dad6b6c7f459a7d2d

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 10, 2018, 23:52

mentioned in commit e2f057605be3b45cfc150e55da660bc25d478b07

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 10, 2018, 23:52

mentioned in commit 0470c56033e51d59397dc27e123c70f55a5bfcdc

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 11, 2018, 24:03

mentioned in commit bdf3dcc15b11afe95a3328b4b8eb69099d0731a9

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 11, 2018, 24:03

mentioned in commit a0b07d9d5e0dbb8163e8818419e153bc13710fd6

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Nov 27, 2018, 17:18

closed via commit d4da27e9abeba6ded2d968b9b65ece15c0ef4c5d
